Payment Term Discounts

agritheory opened the pull request:
Pull-Request
- [X ] Have you followed the guidelines in our Contributing document?
- [X ] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?
- [ X] Have you lint your code locally prior to submission?
- [x] Have you successfully run tests with your changes locally?
- [X ] Does your commit message have an explanation for your changes and why you'd like us to include them?
- [x] Docs have been added / updated
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Did you modify the existing test cases? If yes, why?

---

What type of a PR is this? 
- [X] New Feature Submissions, do not merge, for code review and feedback
--- 

- Motivation and Context (What existing problem does the pull request solve):
> It is common practice in the US and other countries for a supplier to offer a payment discount if the remittance is within an accelerated period. The most common example I am aware of is "2/10 Net 30", where the buyer is given a 2% discount. While that may sound trivial, the businesses that are organized enough to take advantage of it are getting an annualized discount of 20.41% in the 2/10 Net 30 example.

from #14472 

- Related Issues: 
https://github.com/frappe/erpnext/issues/4353
https://github.com/frappe/erpnext/issues/14472

- Screenshots (if applicable, remember, a picture tells a thousand words): 
Screen shots and description of the changes are in #14472 above.

From that proposal issue, a few changes have been made:
-[X] JS code has been broken up a bit more (though I am certainly open to making it a more readable, it's not awesome, as-is)
-[x] Trivial linting of legacy code
-[ ] I think the set invoices methods need a better, more integrated design, I'd like some feedback on this specifically, because I think it means [changing the accounts controller ](https://github.com/frappe/erpnext/blob/7048925016c080d736ed81105842c36da8f840d4/erpnext/controllers/accounts_controller.py#L942)and gets into a scope-of-changes- creep issue

Status:
This is not a final version of this code, I need to incorporate these changes into the tests and get some feedback on the final step: setting invoices as paid.

agritheory committed:
WIP on purchase discounts

rmehta commented on erpnext/accounts/doctype/payment_entry/payment_entry.py:
```diff
@@ -912,4 +913,65 @@ def get_paid_amount(dt, dn, party_type, party, account, due_date):
+
+
+@frappe.whitelist()
+def calc_discount(discount_double_check):
```
https://github.com/frappe/erpnext/wiki/Coding-Standards#break-functions--methods-more-than-10-lines-of-code

agritheory replied:
@rmehta This 10 lines guideline: both JS and py? The link doesn't specify if the style guide is the same for both or not.

rmehta replied:
Obviously both

rmehta reviewed (dismissed):
https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist

vjFaLk commented on erpnext/setup/doctype/company/company.json:
```diff
@@ -1,2888 +1,2987 @@
 {
- "allow_copy": 0,
```
Fix formatting

vjFaLk commented on erpnext/agriculture/doctype/crop_cycle/crop_cycle.js:
```diff
@@ -1,48 +0,0 @@
-// Copyright (c) 2017, Frappe Technologies Pvt. Ltd. and contributors
```
Accidentally deleted this file

vjFaLk commented on erpnext/accounts/doctype/payment_entry/payment_entry.py:
```diff
@@ -912,4 +913,65 @@ def get_paid_amount(dt, dn, party_type, party, account, due_date):
+def set_paid_invoices(transactions):
+	frappe.throw("set INvoices")
+	for item in transactions:
+		frappe.db.set_value(item["reference_doctype"], item["reference_document"], "outstanding_amount", 0)
```
I think there's a way to db_set multiple values at once. Or just do a `get_doc`

vjFaLk commented on erpnext/accounts/doctype/payment_entry/payment_entry.py:
```diff
@@ -912,4 +913,65 @@ def get_paid_amount(dt, dn, party_type, party, account, due_date):
+
+@frappe.whitelist
+def set_paid_invoices(transactions):
+	frappe.throw("set INvoices")
```
Bruh 💩

vjFaLk commented on erpnext/accounts/doctype/payment_terms_template/payment_terms_template.js:
```diff
@@ -8,5 +8,22 @@ frappe.ui.form.on('Payment Terms Template', {
+			if(frm.doc.terms[i]["discount_percent"] < 0){
+				frappe.throw("Please enter the discount as a positive number.");
+			}
+			if(frm.doc.terms[i]["discount_percent"] == 0){
```
Change to `else if`

vjFaLk commented on erpnext/accounts/doctype/payment_entry/payment_entry.py:
```diff
@@ -912,4 +913,65 @@ def get_paid_amount(dt, dn, party_type, party, account, due_date):
+	deductions = []
+	if isinstance(discount_double_check, basestring):
+		discount_double_check = json.loads(discount_double_check)
+	for i in discount_double_check:
```
`i` is a Bad variable name

agritheory commented:
7/11/18 update: I've been refactoring this code into isolated chunks in the 10-line-size. Since this was a originally a custom script too much of the code base was in JS and not enough on the server. I'm working that out now.

shreyashah115 committed:
Merge accounts functionality (#14993)

* Add dialog to merge accounts

* Add conditions and merge functionality

* Fix travis

* Add test case

* Add more test scenarios

agritheory commented:
Test modified is in payment terms template to account for the additional columns relating to discount.

aureliamaeve committed:
Consolidated changes into one commit for review

rohitwaghchaure committed:
Account no + account name in trial balance and financial statements reports (#15038)

rohitwaghchaure committed:
Removed academic term from validation to check duplicate program enrollement (#15046)

agritheory committed:
Merge branch 'develop' into agritheory-discounts

agritheory commented:
Thought this was ready to go but there's a Travis error in the test that isn't showing up in my local version. Will update by EOD EDT 8/1/18.

nabinhait committed:
bumped to version 10.1.45

nabinhait committed:
Merge branch 'hotfix'

nabinhait committed:
Fixed Merge Conflict

Commit:
Simplified Chinese Translation Update  (#14992)

* delete the old version to be replaced by a new one

* New version of the translated simplified Chinese

amrabdelkader7 committed:
Staging fixes (#15051)

* Update hooks.py

* modified date for reload_doc (#15027)

rohitwaghchaure committed:
[Fix] Patient Appointment Calendar not showing data properly (#15052)

nabinhait committed:
Patch fixes

nabinhait committed:
Patch fixes

nabinhait committed:
Update buying_settings.json

nabinhait committed:
Additional fields in Item Price (#14540)

* [IMPROVEMENT] Item Price New Fields and Features

* [IMPROVEMENT] Item Price

 - Item Price insert against Qty, Supplier/Customer, Validup/Validfrom
 - Stock Setting to accept Default Price List for Purchase or Sales Item
 - Move Item Section in Item Price

* Fetch default Item Price based on Customer as null

* test fixes

* Code cleanup and fixed the issue of incorrect item price fetched against the customer

* Removed min qty from packing_increment calculation

* Removed auto_update_price_list_rate from the stock settings

* Revert sales_order.json changes

* Removed sales, purchase price list from stock settings because this fields are already available in selling, buying settings

* Removed unnecessory file

* Fixed item price validation code

Zlash65 committed:
[Enhancement] Tax Withholding Category (#15064)

* add single, cumulative threshold, remove checkboxes

* remove tds child table from supplier & add tds link field

* add description field in tax withholding category

* add tax withholding data for indian setup, some fixes

* add a checkbox for tax withholding in purchase invoice

* remove supplier's child table for tds

* enable tds field if supplier has tds set inits master

* move rates data to child table - adding fiscal year support

* change bootstrap data according to child table config of tds

* show category name in list view

* loyalty program fixes

* moved tax calculation to tax_withholding.py
- calculation for tds amount for cumulative threshold from gl entry

* add fiscal year dependency in company test

* minor loyalty program fix

* minor tier calculation fix

* minor handling duplicate exception

* toggle apply_tds according to supplier, code rectify

* minor fixes for loyalty program

* test case for single and cumulative threshold

* codacy fix

rohitwaghchaure committed:
[Fix] Not able to submit the expense claim (#15057)

rkurungadam committed:
Delivery by Serial No (#15030)

* fields added for delivery by Serial No

* SO - validate item for delivery by Serial No

* Stock Entry - add Serial No on production, validate reserved consumption

* add item by reservation to transaction if delivery by Serial No

* SLE - validate reserved Serial No by SO in Delivery Note, Sale Invoice

* Sales Order - validate Ensure Delivery by Serial No

* Serial No -  remove SO ref on cancel

nabinhait committed:
Get Advances and auto allocate (#14970)

rohitwaghchaure committed:
[Fix] In words showing grand total instead of rounded total (#15065)

KanchanChauhan committed:
Delivery Date Change based on Delivery Date change in child table (#15068)

rohitwaghchaure committed:
[Fix] Task email alert on status change not working if we change the task status from the project form

rkurungadam committed:
Tests SO for Serial No based delivery (#15072)

* fix remove SO from Serial No only when voucher SE

* SO tests - delivery by Serial No

danielchalmers committed:
link to license.txt (#15085)

codingCoffee committed:
Merge pull request #15078 from rohitwaghchaure/task_email_alert_not_working_from_project

Fix] Task's email alert on status change not working if we change the task status from the project form

pratu16x7 committed:
prevent jinja exec in search title

pratu16x7 committed:
Merge pull request #15094 from frappe/search-injection-fix

prevent jinja exec in search title

KanchanChauhan committed:
Get item defaults from item group (#15089)

rohitwaghchaure committed:
[Fix] Incorrect tax amount calculation because of same fieldname in the code

rohitwaghchaure committed:
[Fix] Inactive student validation issue while disabling the student group (#15095)

jigneshpshah committed:
Updated Spelling Maxiumm to Maximum (#15090)

jigneshpshah committed:
Spelling updated from Maxiumm to Maximum (#15091)

jigneshpshah committed:
Spelling updated from Maxiumm to Maximum (#15092)

Spelling updated from Maxiumm to Maximum

rohitwaghchaure committed:
Merge pull request #15097 from rohitwaghchaure/incorrect_tax_amount_in_report

[Fix] Incorrect tax amount calculation because of same fieldname in the code

jigneshpshah committed:
Spelling updated from Maxiumm to Maximum (#15093)

Spelling updated from Maxiumm to Maximum

chdecultot committed:
Correction of the get_price method for pricing rules (#15082)

* Addition of currency arg in get_price

* Change Price to Rate

shreyashah115 committed:
Accounts receivable aging based on supplier due date (#15080)

* Accounts receivable aging based on supplier due date (#13801)

* Remove spaces

shreyashah115 committed:
Force delete item on deletion of Lab test template (#15079)

creamdory committed:
Allow Disabling Batch and edit Validations (#15076)

Zlash65 committed:
add tax withholding & loyalty program in explore view (#15074)

netchampfaris committed:
Use frappe.utils.md_to_html instead of manual import (#15087)

saurabh6790 committed:
bumped to version 10.1.46

saurabh6790 committed:
Merge branch 'hotfix'

saurabh6790 committed:
Resolved merge conflicts

Meritsystems committed:
Point of sale will list only items that are in stock based on option "Display Items in stock" in POS Profle. (#14895)

* Update point_of_sale.py

* Update point_of_sale.py

Add custom field(Display Items In Stock) to POS Profile Doctype of type Check.

* New field " Display Items in Stock" added

* Update point_of_sale.py

* Update point_of_sale.py

ORM used instead of SQL
used Bin to find actual quantity balances.

* Update point_of_sale.py

* Update point_of_sale.py

* Update point_of_sale.py

* Update point_of_sale.py

* Update point_of_sale.py

codingCoffee committed:
[fix] Split issue

take reference name from communication

codingCoffee committed:
Merge pull request #15100 from codingCoffee/splitissue2

[fix] Split issue

rohitwaghchaure committed:
[Fix] 'ascii' codec can't encode error in balance sheet report

rohitwaghchaure committed:
Merge pull request #15101 from rohitwaghchaure/ascii_issue_in_balance_sheet

[Fix] 'ascii' codec can't encode error in balance sheet report

Solufyin committed:
Item Alternate Image Features in e-commerce site. (#15044)

* [ADD] Added Item Alternate Image Features in e-commerce site.

* [IMP] Improved Code for Item Alternate Image Feature.

* [IMP] Improved code for Item alternative image functionality.

* Remove .css file and make a build for erpnext-web.css

* Cleanup styling for item alternate image, also add delegated handler

* Spaces -> Tabs

* Spaces -> Tabs in item.html

ahmedzaqout committed:
fix throw_overlap_error in Shift Assignment (#15103)

fix throw_overlap_error in Shift Assignment

nabinhait committed:
item defaults based on caching

nabinhait committed:
[optimize] Get defaults from item group based on doc

rmehta committed:
fix(style): modal for website

rmehta committed:
optimize(various)

sunnyakaxd committed:
[optimize]: bin updates (#15108)

sunnyakaxd committed:
Cached calls (#15109)

* [optimize]: bin updates

* [optimise]: use frappe.get_cached_value instead of get_value for item

* [fix]: change db.get_cached_value to get_cached_value

nabinhait committed:
[optimize] tree conditions for pricing rule

nabinhait committed:
optimizing transactions by caching

rohitwaghchaure committed:
[Fix] Math domain error (#15107)

nabinhait committed:
minor fix

nabinhait committed:
minor fix

rohitwaghchaure committed:
[Fix] Not able to delete the task even if task removed from the project (#15105)

* [Fix] Not able to delete the task even if task removed from the project

* [Fix] Custom field's data in task not updating from the project

rohitwaghchaure committed:
[Fix] disable_rounded_total not found error (#15113)

rohitwaghchaure committed:
[Minor] Added condition for get_bank_cash_account (#15112)

nabinhait committed:
minor fix

nabinhait committed:
minor fix

nabinhait committed:
minor fixes

saurabh6790 committed:
bumped to version 10.1.47

saurabh6790 committed:
Merge branch 'hotfix'

saurabh6790 committed:
[Resolved] merge conflicts

nabinhait committed:
fix(various)

agritheory commented:
I've merged with the latest versions of Frappe and ERPnext locally and cannot replicate any of the conflicts. Running test as `bench run-tests --doctype "Payment Entry"` on a fresh site yields no errors.
@vjFaLk or @rmehta if you could resume code review and/or mark for merge I think this is ready.

aureliamaeve committed:
Merge branch 'agritheory-discounts' of https://github.com/agritheory/erpnext into agritheory-discounts

aureliamaeve committed:
Continue testing locally in Payment Term Template

agritheory commented:
@vjFaLk Travis is now passing. The codacy requests don't seem worth another commit. Please advise.

vjFaLk commented:
Yesus, this is a commit mess. Will need to do a lot of git wizardry. Lemme pull your branch and clean it up

heatherpotter commented:
Please don't go stale! I really really need this!

vjFaLk commented:
@nabinhait sir pls

FinForceConsulting commented:
@agritheory, did you create a new PR for this? please share link

gdlrsilv3 commented:
hi @agritheory  , were you able to add the payment term discounts functionality to ERP Next?
